lib/opal/nodes/literal.rb

```diff
@@ -213,42 +213,132 @@ class XStringNode < Base
       handle :xstr
 
       def compile
-        children.each do |child|
-          case child.type
-          when :str
-            value = child.loc.expression.source
-            if expr? && children.size == 1
-              value = detect_and_remove_trailing_semicolon(value, child)
+        should_add_semicolon = false
+        has_embeded_return   = false
+        returning, children  = unpack_return(self.children)
```

Comment:
What do you think about moving children dispatching to constructor? (like we do in `def/iter/call` nodes)